pex/tracer.py

```diff
@@ -8,29 +8,36 @@
 import time
 from contextlib import contextmanager
 
+from pex.typing import TYPE_CHECKING, cast
 from pex.variables import ENV
 
 __all__ = ("TraceLogger",)
 
+if TYPE_CHECKING:
+    from typing import Any, Callable, IO, Iterator, List, Optional
+
 
 class Trace(object):
     __slots__ = ("msg", "verbosity", "parent", "children", "_clock", "_start", "_stop")
 
     def __init__(self, msg, parent=None, verbosity=1, clock=time):
+        # type: (str, Optional[Trace], int, Any) -> None
         self.msg = msg
         self.verbosity = verbosity
         self.parent = parent
         if parent is not None:
-            parent.children.append(self)
-        self.children = []
+            parent.children.append(self)  # type: ignore[has-type]
```

Comment:
I couldn't get MyPy happy with this, even with using `cast(Trace, parent)` and `cast("List[Trace]", parent.children)`. It keeps complaining `Cannot determine type of 'children'  [has-type]`.

Reply:
Perhaps the idiom note below clears this up. If not, perhaps an isinstance check instead of a not None check. Hopefully the former.